File: python/pyjse/sql.py
"""SQL generation for JSE query patterns."""

QUERY_FIELDS = "subject, predicate, object, meta"


def pattern_to_triple(subject: str, predicate: str, object_: str) -> list:
    """Convert $pattern args to triple for PostgreSQL jsonb containment."""
    if subject == "$*" and object_ == "$*":
        return [predicate]
    return [subject, predicate, object_]


def triple_to_sql_condition(triple: list) -> str:
    """Build SQL WHERE clause for a triple pattern."""
    import json

    doc = json.dumps({"triple": triple})
    escaped = doc.replace("'", "''")
    return f"meta @> '{escaped}'"

File: typescript/src/sql.ts
/**
 * Query field list for SQL SELECT.
 */
export const QUERY_FIELDS =
  "subject, predicate, object, meta";

/**
 * Convert $pattern arguments to PostgreSQL jsonb containment triple.
 * - ["$pattern", "$*", "author of", "$*"] -> triple: ["author of"]
 * - ["$pattern", "Liu Xin", "author of", "$*"] -> triple: ["Liu Xin", "author of", "$*"]
 */
export function patternToTriple(
  subject: string,
  predicate: string,
  object: string
): unknown[] {
  if (subject === "$*" && object === "$*") {
    return [predicate];
  }
  return [subject, predicate, object];
}

/**
 * Build SQL WHERE clause for a triple pattern.
 */
export function tripleToSqlCondition(triple: unknown[]): string {
  const json = JSON.stringify({ triple });
  const escaped = json.replace(/'/g, "''");
  return `meta @> '${escaped}'`;
}
